### PR TITLE
Stake Delegation

### PR DESCRIPTION
**Pull request by Ramarti**
This is PR adds stake allocation and delegation, includes major refactors, in part forced by the size limit of `FortaStaking`, and in part by trying to have a more clear architecture, plus going over implications of delegation over other contracts.

It's looking good though.

# Refactor of FortaStaking
- Improved folder structure
- Stake subject interface granularity. We now have `DirectStakeSubjects`, with their own stake thresholds (AGENTS, old SCANNERS) and `DelegatedStakeSubjects`, with managed stake thresholds.
- `FortaStakingParameters` renamed to `StakeSubjectHandler`.
- `StakeSubjectHandler` expands to accomodate managed subject values, and the lack of intrinsic stake thresholds for DELEGATOR and DELEGATED
- `StakeAllocator` is introduced to manage the allocation of stake for DELEGATOR and DELEGATED.
- Slashing goes to a folder too
- Slashing propagates to delegators by a percentage

# Slashing
- Added limitations to not slash directly delegators

# TODO
- Fix StakingEscrow tests
- Rewards
- Stake migration

### CODE REVIEW

File: contracts/components/staking/stakeSubjectHandling/IStakeSubjectHandler.sol [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+pragma solidity ^0.8.9;
+
+import "./IStakeSubject.sol";
+
+interface IStakeSubjectHandler {
+    event StakeSubjectChanged(address newHandler, address oldHandler);
+    function setStakeSubject(uint8 subjectType, address subject) external;
+    function getStakeSubject(uint8 subjectType) external view returns (address);
+    function activeStakeFor(uint8 subjectType, uint256 subject) external view returns (uint256);
+    function maxStakeFor(uint8 subjectType, uint256 subject) external view returns (uint256);
```
Comment (frangio) [inline contracts/components/staking/stakeSubjectHandling/IStakeSubjectHandler.sol:13]: The name `subject` is used for `address subject` in `setStakeSubjet` and also for `uint subject` in `maxStakeFor` (etc.) but I believe they refer to completely different things.

In general I think "subject" and related concepts are quite overloaded, and it makes reading the code difficult.

In particular, note that in this PR there was this series of renames:
- `IStakeController` -> `IStakeSubjectHandler`
- `setStakeSubjectHandler` -> `setStakeSubject`

The concept of "stake subject handler" was reused for the interface name, but for almost the opposite thing. Note that the parameters of the event `StakeSubjectChanged` are still using the "handler" term!

---

The way I understand all this:
- A "subject" is an individual node runner, scanner, or agent.
  - Each of these individuals is a "stake subject" in the sense that they can have stake put on them.
- Each individual has a "subject id" that identifies them, which are respectively an integer (sequential), an address, and an integer.
  - What is called `subject` in `maxStakeFor` and all the other getters could be called `subjectId`, but this distinction is not so important (because a subject is always represented by its id).
- There are "subject registries", respectively NodeRunnerRegistry, ScannerRegistry (being phased out), and AgentRegistry. The subject registries contain various data about the subjects themselves, and in particular about their stake thresholds and allocations (although the stake itself is elsewhere, in the FortaStaking contract).
  - **Suggestion:** `setStakeSubject(..., address subject)` should be called or `setStakeSubjectRegistry(..., address registry)` (or something involving "registry").
- There is a central contract that answers queries about stake and stake-related data, by directing them to the appropriate registry.
  - This is currently the so-called "stake subject handler", but as noted above this term seemed to refer to the registries previously.
  - **Suggestion:** `StakeSubjectHandler` needs a different name. IMO even `StakeController` is better, but I can also suggest `StakeSubjectQueries`. "Handler" is not right, this contract is not handling, it is directing queries to the appropriate registry.

Comment (Ramarti) [reply contracts/components/staking/stakeSubjectHandling/IStakeSubjectHandler.sol:13]: Apologies for the inconsistencies, I might have left some stray "handlers".

Minor correction: IStakeController -> IStakeSubjectHandle should be FortaStakingParameters -> IStakeSubjectHandler

> Each individual has a "subject id" that identifies them, which are respectively an integer (sequential), an address, and an integer.

It's all integers. In the case of ScannerSubject it is still an integer, converted from address

> What is called subject in maxStakeFor and all the other getters could be called subjectId, but this distinction is not so important (because a subject is always represented by its id).

unification might be important, but not sure if I have time to go throught that one.

> Suggestion: setStakeSubject(..., address subject) should be called or setStakeSubjectRegistry(..., address registry) (or something involving "registry").

Why not

> Suggestion: StakeSubjectHandler needs a different name. IMO even StakeController is better, but I can also suggest StakeSubjectQueries. "Handler" is not right, this contract is not handling, it is directing queries to the appropriate registry.

Controller is very generic too, since it is not controlling anything. `StakeSubjectRepository` sounds better IMO, what do you think?

Comment (Ramarti) [reply contracts/components/staking/stakeSubjectHandling/IStakeSubjectHandler.sol:13]: StakeSubjectGateway it is

File: contracts/components/agents/AgentRegistryCore.sol [hunk 2]
```diff
@@ -118,17 +118,18 @@ abstract contract AgentRegistryCore is
         return _stakeThreshold;
     }
 
+    function _isStakedActivated() internal view returns(bool) {
```
[the remark this answers is not in the document]
Comment (Ramarti) [reply contracts/components/agents/AgentRegistryCore.sol:121]: Why? we are using _ prefix for internal and private. It is used only in `AgentRegistryEnable`

File: contracts/components/staking/slashing/SlashingController.sol [hunk 3]
```diff
@@ -75,7 +75,7 @@ contract SlashingController is BaseComponentUpgradeable, StateMachineController,
     );
     event EvidenceSubmitted(uint256 proposalId, StateMachines.State stateId, string[] evidence);
     event SlashingExecutorChanged(address indexed slashingExecutor);
-    event StakingParametersManagerChanged(address indexed stakingParametersManager);
+    event subjectHandlerChanged(address indexed subjectHandler);
```
[the remark this answers is not in the document]
Comment (Ramarti) [reply contracts/components/staking/slashing/SlashingController.sol:78]: Find & Replace, not even once

**Comment by frangio** [comment]
The new directory is called `stakeSubjectHandling` using camel case. There is a preexisting directory using snake case `node_runners`. Let's use one style consistely. (For directories, I personally always use dashes as word separators.)

File: contracts/components/scanners/ScannerToNodeRunnerMigration.sol [hunk 4]
```diff
@@ -122,5 +131,21 @@ contract ScannerToNodeRunnerMigration is BaseComponentUpgradeable {
         return nodeRunnerId;
     }
 
+    function _getNodeRunnerIdOrMint(address nodeRunner, uint256 nodeRunnerId, uint256 chainId) private returns(uint256) {
+        if (nodeRunnerId == NODE_RUNNER_NOT_MIGRATED) {
+            if (_migratedNodeRunners[chainId][nodeRunner] != 0) {
+                revert NodeRunnerAlreadyMigrated(_migratedNodeRunners[chainId][nodeRunner]);
```
Comment (frangio) [inline contracts/components/scanners/ScannerToNodeRunnerMigration.sol:137]: Why not just return the existing id?

File: contracts/components/scanners/ScannerToNodeRunnerMigration.sol [hunk 5]
```diff
@@ -80,19 +85,21 @@ contract ScannerToNodeRunnerMigration is BaseComponentUpgradeable {
...
+        uint256 chainId
     ) private returns (uint256) {
-        uint256 nodeRunnerId = inputNodeRunnerId;
+        uint256 nodeRunnerId = _getNodeRunnerIdOrMint(nodeRunner, inputNodeRunnerId, chainId);
         if (nodeRunnerRegistry.balanceOf(nodeRunner) == 0 && nodeRunnerId == NODE_RUNNER_NOT_MIGRATED) {
-            nodeRunnerId = nodeRunnerRegistry.registerMigratedNodeRunner(nodeRunner);
+            nodeRunnerId = nodeRunnerRegistry.registerMigratedNodeRunner(nodeRunner, chainId);
         } else if (nodeRunnerRegistry.ownerOf(nodeRunnerId) != nodeRunner) {
             revert NotOwnerOfNodeRunner(nodeRunner, nodeRunnerId);
         }
```
Comment (frangio) [inline contracts/components/scanners/ScannerToNodeRunnerMigration.sol:105]: This should all be removed now. It's part of `_getNodeRunnerIdOrMint`.

```suggestion
```

File: contracts/components/node_runners/NodeRunnerRegistryCore.sol [hunk 6]
```diff
@@ -349,42 +362,34 @@ abstract contract NodeRunnerRegistryCore is BaseComponentUpgradeable, ERC721Upgr
...
+        if (chainId == 0) revert ZeroAmount("chainId");
         if (newStakeThreshold.max <= newStakeThreshold.min) revert StakeThresholdMaxLessOrEqualMin();
-        emit StakeThresholdChanged(newStakeThreshold.min, newStakeThreshold.max, newStakeThreshold.activated);
-        _stakeThreshold = newStakeThreshold;
+        emit ManagedStakeThresholdChanged(chainId, newStakeThreshold.min, newStakeThreshold.max, newStakeThreshold.activated);
+        _scannerStakeThresholds[chainId] = newStakeThreshold;
     }
 
     /**
      * @notice Getter for StakeThreshold for the scanner with id `subject`
```
Comment (frangio) [inline contracts/components/node_runners/NodeRunnerRegistryCore.sol:384]:
```suggestion
     * @notice Getter for StakeThreshold for the scanner with id `managedId`
```

File: contracts/components/staking/stakeSubjectHandling/StakeSubjectHandler.sol [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+// See Forta Network License: https://github.com/forta-network/forta-contracts/blob/master/LICENSE.md
+
+pragma solidity ^0.8.9;
+
+import "../FortaStaking.sol";
+import "./IDelegatedStakeSubject.sol";
+import "./IDirectStakeSubject.sol";
+
+/**
+ * Formerly StakeSubjectHandler.
```
Comment (frangio) [inline contracts/components/staking/stakeSubjectHandling/StakeSubjectHandler.sol:11]: Some search & replace went wrong here.

File: contracts/components/staking/SubjectTypeValidator.sol [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+            revert InvalidSubjectType(subjectType);
+        _;
+    }
+
+    modifier onlyAgencyType(uint8 subjectType, SubjectStakeAgency expected) {
+        if (getSubjectTypeAgency(subjectType) != expected) revert ForbiddenForType(subjectType, getSubjectTypeAgency(subjectType), expected);
+        _;
+    }
+
+    modifier notAgencyType(uint8 subjectType, SubjectStakeAgency forbidden) {
```
Comment (frangio) [inline contracts/components/staking/SubjectTypeValidator.sol:54]: I'm wary of this modifier and in general restricting something to "not X" because the set that satisfies "not X" may grow in the future to include new values (new agency types in this case). Where would this be used? I would prefer to use `only`-type modifiers exclusively.

File: contracts/components/staking/SubjectTypeValidator.sol [hunk 9]
```diff
@@ -0,0 +1,85 @@
+// SPDX-License-Identifier: UNLICENSED
```
Comment (frangio) [inline contracts/components/staking/SubjectTypeValidator.sol:1]: We need to delete the file `SubjectTypes.sol` which apparently was moved to this one but got left over in some merge.

File: contracts/components/staking/SubjectTypeValidator.sol [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+// See Forta Network License: https://github.com/forta-network/forta-contracts/blob/master/LICENSE.md
+
+pragma solidity ^0.8.9;
+
+uint8 constant UNDEFINED_SUBJECT = 255;
+uint8 constant SCANNER_SUBJECT = 0;
+uint8 constant AGENT_SUBJECT = 1;
+uint8 constant NODE_RUNNER_SUBJECT = 2;
+uint8 constant DELEGATOR_NODE_RUNNER_SUBJECT = 3;
```
Comment (frangio) [inline contracts/components/staking/SubjectTypeValidator.sol:10]: I'm confused by this distinction. What is a "delegator node runner"?

And it seems we changed the value of `NODE_RUNNER_SUBJECT`, it used to be 3 and is now 2. Is this ok?

File: contracts/components/staking/FortaStaking.sol [hunk 11]
```diff
@@ -353,6 +353,9 @@ contract FortaStaking is BaseComponentUpgradeable, ERC1155SupplyUpgradeable, Sub
         uint256 stakeValue = activeSharesToStake(activeSharesId, activeShares);
         uint256 inactiveShares = stakeToInactiveShares(FortaStakingUtils.activeToInactive(activeSharesId), stakeValue);
 
+        if (getSubjectTypeAgency(subjectType) == SubjectStakeAgency.DELEGATED || getSubjectTypeAgency(subjectType) == SubjectStakeAgency.DELEGATOR) {
```
Comment (frangio) [inline contracts/components/staking/FortaStaking.sol:356]: Please define a variable for `getSubjectTypeAgency(subjectType)`.

File: contracts/components/staking/FortaStaking.sol [hunk 12]
```diff
@@ -507,11 +551,12 @@ contract FortaStaking is BaseComponentUpgradeable, ERC1155SupplyUpgradeable, Sub
      * @notice Release reward owed by given `account` for its current or past share for a given `subject`.
      * @dev If staking from a contract, said contract may optionally implement ERC165 for IRewardReceiver.
      * Emits a Release event.
-     * @param subjectType agents, scanner or future types of stake subject. See SubjectTypes.sol
+     * @param subjectType type id of Stake Subject. See SubjectTypeValidator.sol
      * @param subject id identifying subject (external to FortaStaking).
      * @param account that staked on the subject.
      * @return available reward transferred.
      */
+    /*
```
Comment (frangio) [inline contracts/components/staking/FortaStaking.sol:559]: Let's delete all the code we're removing, instead of commenting it out.

File: contracts/components/staking/FortaStaking.sol [hunk 13]
```diff
@@ -104,9 +102,10 @@ contract FortaStaking is BaseComponentUpgradeable, ERC1155SupplyUpgradeable, Sub
     event Released(uint8 indexed subjectType, uint256 indexed subject, address indexed to, uint256 value);
     event DelaySet(uint256 newWithdrawalDelay);
     event TreasurySet(address newTreasury);
-    event StakeParamsManagerSet(address indexed newManager);
+    event StakeHelpersConfigured(address indexed subjectManager, address indexed allocator);
```
Comment (frangio) [inline contracts/components/staking/FortaStaking.sol:105]: Please note it says "manager" here and I think it will have to be renamed to "gateway".

File: contracts/components/staking/FortaStaking.sol [hunk 14]
```diff
@@ -697,10 +747,17 @@ contract FortaStaking is BaseComponentUpgradeable, ERC1155SupplyUpgradeable, Sub
     }
 
     // Admin: change staking parameters manager
-    function setStakingParametersManager(IStakeController newStakingParameters) public onlyRole(DEFAULT_ADMIN_ROLE) {
-        if (address(newStakingParameters) == address(0)) revert ZeroAddress("newStakingParameters");
-        emit StakeParamsManagerSet(address(newStakingParameters));
-        _stakingParameters = newStakingParameters;
+    function configureStakingHelpers(IStakeSubjectHandler __subjectHandler, IStakeAllocator __allocator) public onlyRole(DEFAULT_ADMIN_ROLE) {
```
Comment (frangio) [inline contracts/components/staking/FortaStaking.sol:750]: Function name is "configure **staking** helpers" and event name is "**stake** helpers configured".

Let's settle on Stake.

File: contracts/components/staking/FortaStaking.sol [hunk 15]
```diff
@@ -397,58 +400,98 @@ contract FortaStaking is BaseComponentUpgradeable, ERC1155SupplyUpgradeable, Sub
...
         // See deposit and stakeToActiveShares methods.
-        uint256 maxSlashableStake = Math.mulDiv(activeStake + inactiveStake, _stakingParameters.maxSlashableStakePercent(), HUNDRED_PERCENT);
+        uint256 maxSlashableStake = Math.mulDiv(activeStake + inactiveStake, MAX_SLASHABLE_PERCENT, HUNDRED_PERCENT);
+
         if (stakeValue > maxSlashableStake) revert SlashingOver90Percent();
 
-        uint256 slashFromActive = Math.mulDiv(activeStake, stakeValue, activeStake + inactiveStake);
+        slashFromActive = Math.mulDiv(activeStake, stakeValue, activeStake + inactiveStake);
         uint256 slashFromInactive = stakeValue - slashFromActive;
-        stakeValue = slashFromActive + slashFromInactive;
```
Comment (frangio) [inline contracts/components/staking/FortaStaking.sol:427]: Why was this line deleted or not moved to the `slash` function?

File: contracts/components/agents/AgentRegistryCore.sol (hunk 2 above)
[the remark this answers is not in the document]
Comment (Ramarti) [reply contracts/components/agents/AgentRegistryCore.sol:121]: Oh right, didn't notice.

File: contracts/vesting/escrow/StakingEscrow.sol [hunk 16]
```diff
@@ -143,12 +143,14 @@ contract StakingEscrow is Initializable, ERC165, IRewardReceiver, ForwardedConte
...
      * through this relay function.
-     * @param subjectType agents, scanner or future types of stake subject. See SubjectTypes.sol
+     * @param subjectType agents, scanner or future types of stake subject. See SubjectTypeValidator.sol
      * @param subject id identifying subject (external to FortaStaking).
      * @return released amount
      */
     function claimReward(uint8 subjectType, uint256 subject) public returns (uint256) {
-        return l2staking.releaseReward(subjectType, subject, address(this));
+        // return l2staking.releaseReward(subjectType, subject, address(this));
+        // TODO: fix
```
[the remark this answers is not in the document]
Comment (Ramarti) [reply contracts/vesting/escrow/StakingEscrow.sol:152]: Nope

File: contracts/components/node_runners/NodeRunnerRegistryCore.sol [hunk 17]
```diff
@@ -174,18 +185,7 @@ abstract contract NodeRunnerRegistryCore is BaseComponentUpgradeable, ERC721Upgr
...
-                            req.scanner,
-                            req.nodeRunnerId,
-                            req.chainId,
-                            keccak256(
-                                abi.encodePacked(req.metadata)
-                            ),
-                            req.timestamp
-                        )
-                    )
+                    keccak256(abi.encode(_SCANNERNODEREGISTRATION_TYPEHASH, req.scanner, req.nodeRunnerId, req.chainId, keccak256(abi.encodePacked(req.metadata)), req.timestamp))
```
[the remark this answers is not in the document]
Comment (Ramarti) [reply contracts/components/node_runners/NodeRunnerRegistryCore.sol:188]: Is there a standard / widely adopted number or is it just a subjective appreciation?

File: contracts/components/node_runners/NodeRunnerRegistryCore.sol [hunk 18]
```diff
@@ -280,11 +276,16 @@ abstract contract NodeRunnerRegistryCore is BaseComponentUpgradeable, ERC721Upgr
...
-        // _isStakedOverMin already checks for disabled, but returns true in every case if stakeController is not set.
-        // since isStakedOverMin() is external, we need to keep this duplicate check.
-        return _scannerNodes[scanner].registered &&
-            !_scannerNodes[scanner].disabled; // && TODO: stake
-            // getStakeController().getStakeSubjectHandler(SCANNER_SUBJECT).isStakedOverMin(scannerAddressToId(scanner));
+        bool result = _scannerNodes[scanner].registered && !_scannerNodes[scanner].disabled;
+        if (_scannerStakeThresholds[_scannerNodes[scanner].chainId].activated) {
+            result = result && _isScannerStakedOverMin(scanner);
+        }
+        return result && _exists(_scannerNodes[scanner].nodeRunnerId);
```
[the remark this answers is not in the document]
Comment (Ramarti) [reply contracts/components/node_runners/NodeRunnerRegistryCore.sol:283]: Oh I right, those are pointers. Meaning they don't allocate an extra variable. This is going to make stuff more legible going forward, thanks.

File: contracts/components/staking/FortaStaking.sol (hunk 13 above)
Comment (Ramarti) [reply contracts/components/staking/FortaStaking.sol:105]: Noted

File: contracts/components/staking/FortaStaking.sol (hunk 15 above)
Comment (Ramarti) [reply contracts/components/staking/FortaStaking.sol:427]: cause it is not really necessary. I think it's a left over on a comment I have, `stakeValue` is defined earlier, and it _is_ `slashFromActive + slashFromInactive`, since those are derived from stakeValue.

File: contracts/components/staking/stakeSubjectHandling/StakeSubjectHandler.sol [hunk 19]
```diff
@@ -0,0 +1,171 @@
...
+     * @param __fortaStaking address of FortaStaking.
+     */
+    function initialize(address __manager, address __fortaStaking) public initializer {
+        __BaseComponentUpgradeable_init(__manager);
+
+        _setFortaStaking(__fortaStaking);
+    }
+
+    /// Setter for FortaStaking implementation address.
+    function setFortaStaking(address newFortaStaking) external onlyRole(DEFAULT_ADMIN_ROLE) {
```
[the remark this answers is not in the document]
Comment (Ramarti) [reply contracts/components/staking/stakeSubjectHandling/StakeSubjectHandler.sol:42]: Will delete

File: contracts/components/staking/FortaStaking.sol (hunk 14 above)
Comment (Ramarti) [reply contracts/components/staking/FortaStaking.sol]: bea9f56

File: contracts/components/staking/FortaStaking.sol (hunk 13 above)
Comment (Ramarti) [reply contracts/components/staking/FortaStaking.sol]: And fixed

File: contracts/components/staking/FortaStaking.sol (hunk 11 above)
Comment (Ramarti) [reply contracts/components/staking/FortaStaking.sol]: I defined it in every instance where is at least 2 `getSubjectTypeAgency(subjectType)`

File: contracts/components/staking/FortaStaking.sol (hunk 12 above)
Comment (Ramarti) [reply contracts/components/staking/FortaStaking.sol]: Done

File: contracts/components/node_runners/NodeRunnerRegistryCore.sol (hunk 17 above)
[the remark this answers is not in the document]
Comment (frangio) [reply contracts/components/node_runners/NodeRunnerRegistryCore.sol:188]: I tend to use `120`, but I think `150` was ok too.

File: contracts/components/staking/FortaStaking.sol [hunk 20]
```diff
@@ -397,58 +400,98 @@ contract FortaStaking is BaseComponentUpgradeable, ERC1155SupplyUpgradeable, Sub
...
     function slash(
         uint8 subjectType,
         uint256 subject,
         uint256 stakeValue,
         address proposer,
         uint256 proposerPercent
-    ) public override onlyRole(SLASHER_ROLE) returns (uint256) {
+    ) public override onlyRole(SLASHER_ROLE) notAgencyType(subjectType, SubjectStakeAgency.DELEGATOR) returns (uint256) {
         uint256 activeSharesId = FortaStakingUtils.subjectToActive(subjectType, subject);
+        uint256 slashFromActive = _slash(activeSharesId, subjectType, subject, stakeValue);
```
Comment (frangio) [inline contracts/components/staking/FortaStaking.sol:421]: Shouldn't we be slashing `stakeValue * (100 - SLASH_DELEGATOR_PERCENT) / 100` ?
